userscript-preferences.js

```diff
@@ -0,0 +1,230 @@
+            case "array_checkbox":
+                for ( let j=0; j<pref.options.length; j++ ) {
+                    let key     = pref.options[j].key;
+                    let label   = pref.options[j].label || key;
+                    let checked = ( this_value.hasOwnProperty(key) && this_value[key] ) ? "checked=checked" : "";
+                    dd.append(
+                        "<label>"+
+                        `<input data-userscript-pref='${this_id}' name='${key}' type='checkbox' value=1 ${checked} />`+
+                        label+
+                        "</label>"+
+                        "<br />"
+                    );
+                }
+                break;
+            case "radio":
+                // let item_id = this_id + "_" + i;
+                for ( let j=0; j<pref.options.length; j++ ) {
+                    let value = pref.options[j];
+                    let checked = value === this_value ? "checked=checked" : "";
+                    dd.append(
+                        "<label>"+
+                        `<input data-userscript-pref='${this_id}' name='${this_id}' type='radio' value='${value}' ${checked} />`+
+                        value+
+                        "</label>"+
+                        "<br />"
+                    );
+                }
+                break;
+        }
+    }
+
+    let save = $( `<button data-userscript-save='${def.key}'>Save "${label}" Settings</button>` );
+    save.click( function() {
+        _save_userscript_settings(def);
+    } );
+
+    core_prefs.append(
+        "<dt></dt>",
+        $("<dd></dd>").append( save )
+    );
+}
+
+function _save_userscript_settings(def) {
+    "use strict";
+
+    let values = {};
+
+    for ( let i=0; i<def.options.length; i++ ) {
+        let pref    = def.options[i];
+        let this_id = def.key + "_" + pref.key;
+
+        switch ( pref.type ) {
+            case "text":
+                values[ pref.key ] = $(`input[data-userscript-pref=${this_id}]`).first().val();
+                break;
+            case "checkbox":
+                let is_checked = $(`input[data-userscript-pref=${this_id}]`).first().prop("checked");
+                values[ pref.key ] = is_checked ? true : false;
+                break;
+            case "array_checkbox":
+                let all_checkboxes = $(`input[data-userscript-pref=${this_id}]`);
+                values[ pref.key ] = {};
+                all_checkboxes.map( function() {
+                    values[ pref.key ][ $(this).prop("name") ] = $(this).prop("checked") ? true : false;
+                } );
+                break;
+            case "radio":
+                values[ pref.key ] = $(`input[data-userscript-pref=${this_id}]:checked`).val();
+                break;
+
+        }
+    }
+
+    localStorage.setItem(
+        def.key,
+        JSON.stringify( values )
+    );
+
+    if ( def.hasOwnProperty("callback") ) {
+        def["callback"]( values );
+    }
```

Comment:
1. Shouldn't the callback be triggered before the values / JSON is saved to localStorage? That would give you more options for settings validation (e.g., turning one setting off disables a group of sub-settings / makes them irrelevant).

2. This looks like a top-level callback, whereas the example JSON above shows per-setting callbacks. Did you want to allow individual settings to have their own callback?  I prefer that approach, since it would allow each setting to fully own its own callback stuff (I'm thinking "single-responsibility principle", "open-closed principle", etc.).  (And, FWIW, I did find having a setting-specific callback useful for the existing userscript script I added preferences to.)

Reply:
Would per-setting callbacks just receive the single value for that setting?
That wouldn't allow dependant / multi-setting validation.
I feel a combination of this & pre-store callbacks would greatly complicate things. If we store the value returned by the callback instead of the user-entered value, do we then need to update the UI to show the changed value? Would a user be confused if they enter a value & it then immediately changes?
Can you give your use-case you mentioned?